csharp/ql/src/semmle/code/csharp/PrintAst.qll

```diff
@@ -0,0 +1,606 @@
+  }
+}
+
+/**
+ * A node representing an `Attribute`.
+ */
+final class AttributeNode extends AstNode {
+  Attribute attr;
+
+  AttributeNode() {
+    attr = ast and
+    not isCompilerGeneratedAttributable(attr.getTarget()) and
+    not isInsideUnneededAttributable(attr.getTarget())
+  }
+
+  override AstNode getChild(int childIndex) { result.getAst() = attr.getChild(childIndex) }
+}
+
+/**
+ * A node representing a `TypeParameter`.
+ */
+final class TypeParameterNode extends AstNode {
+  TypeParameter typeParameter;
+
+  TypeParameterNode() {
+    typeParameter = ast and
+    not isInsideUnneededUnboundGeneric(typeParameter.getDeclaringGeneric())
+  }
+
+  override AstNode getChild(int childIndex) { none() }
+}
+
+/**
+ * A node representing a `ValueOrRefType`.
+ */
+final class TypeNode extends AstNode {
+  ValueOrRefType type;
+
+  TypeNode() {
+    type = ast and
+    not type instanceof TupleType and
+    not type instanceof ArrayType and
+    not type instanceof NullableType and
+    not isInsideUnneededType(type)
+  }
+
+  override PrintAstNode getChild(int childIndex) {
+    childIndex = 0 and
+    result.(AttributesNode).getAttributable() = type
+    or
+    childIndex = attributableOffset(type) and
+    result.(TypeParametersNode).getUnboundGeneric() = type
+    or
+    childIndex = attributableOffset(type) + unboundGenericOffset(type) and
+    result.(ParametersNode).getParameterizable() = type
+    or
+    result.(AstNode).getAst() =
+      rank[childIndex - attributableOffset(type) - unboundGenericOffset(type) -
+          parameterizableTypeOffset(type)](Member m, string file, int line, int column |
+        m = type.getAMember() and
+        locationSortKeys(m, file, line, column)
+      |
+        m order by file, line, column
+      )
+  }
+}
+
+/**
+ * A node representing a `NamespaceDeclaration`.
+ */
+final class NamespaceNode extends AstNode {
+  NamespaceDeclaration namespace;
+
+  NamespaceNode() { namespace = ast }
+
+  override PrintAstNode getChild(int childIndex) {
+    result.(AstNode).getAst() =
+      rank[childIndex](Element a, string file, int line, int column |
+        (a = namespace.getAChildNamespaceDeclaration() or a = namespace.getATypeDeclaration()) and
+        locationSortKeys(a, file, line, column)
+      |
+        a order by file, line, column
+      )
+  }
+}
+
+/**
+ * A node representing the parameters of a `Parameterizable`.
+ * Only rendered if there's at least one parameter and if the
+ * `Parameterizable` is not compiler generated or is of type
+ * `Accessor`.
+ */
+final class ParametersNode extends PrintAstNode, TParametersNode {
+  Parameterizable parameterizable;
+
+  ParametersNode() {
+    this = TParametersNode(parameterizable) and
+    parameterizable.getNumberOfParameters() > 0 and
+    not isInsideUnneededParameterizable(parameterizable) and
+    (
+      not parameterizable.isCompilerGenerated() or
+      parameterizable instanceof Accessor
+    )
```

Comment:
You may as well move these constraints into the body of the `TParametersNode` `newtype`-branch.

Reply:
Thanks. Is there any benefit of doing this? Other than having all constraints in one place? 

I guess it is possible to have multiple derived classes of a `newtype`-branch, and then I could have different restrictions on them. Is this a pattern that we use or we typically have one derived type of a `newtype`-branch, and then restrict that further by subtyping the class?

Reply:
The benefit is that the implicit predicate generated by the QL compiler for the `TParametersNode` injector becomes smaller. Of course this only works, as you mention, if all classes that derive from the injector apply the same constraints.

Reply:
I understand most part of what's written here. :-) What is an injector?

Reply:
Sorry, not sure if it is standard terminology, but it is branch in a `newtype`.